WebDriver: [Cocoa] [Actions] [Key] Right-hand modifier keys report as their left-hand counterpart

patrickangle opened the pull request:
#### 9aaabd386a2fd8d1c1689cae9c4d3716ec8cc71a
<pre>
WebDriver: [Cocoa] [Actions] [Key] Right-hand modifier keys report as their left-hand counterpart
<a href="https://bugs.webkit.org/show_bug.cgi?id=248674">https://bugs.webkit.org/show_bug.cgi?id=248674</a>
rdar://100588604

Reviewed by BJ Burg.

Add support to the Cocoa ports for properly handling right-hand modifiers which, when paired with safaridriver changes,
will allow the left and right modifier keys to be distiguished from one another. Most of the support is already done,
this just adds the missing `CommandRight` virtual key to the automation protocol and adds handling so that these keys
are properly considered to be a modifier key when combined with other subsequent actions.

This progresses the following WPT test cases:
webdriver/tests/perform_actions/key_events.py::test_modifier_key_sends_correct_events[\ue052-R_ALT]
webdriver/tests/perform_actions/key_events.py::test_modifier_key_sends_correct_events[\ue051-R_CONTROL]
webdriver/tests/perform_actions/key_events.py::test_modifier_key_sends_correct_events[\ue050-R_SHIFT]
webdriver/tests/perform_actions/key_events.py::test_special_key_sends_keydown[R_ALT-expected48]
webdriver/tests/perform_actions/key_events.py::test_special_key_sends_keydown[R_CONTROL-expected53]
webdriver/tests/perform_actions/key_events.py::test_special_key_sends_keydown[R_SHIFT-expected61]

* Source/WebKit/UIProcess/Automation/Automation.json:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::normalizedVirtualKey):
* Source/WebKit/UIProcess/Automation/cocoa/WebAutomationSessionCocoa.mm:
(WebKit::WebAutomationSession::charCodeForVirtualKey const):
* Source/WebKit/UIProcess/Automation/ios/WebAutomationSessionIOS.mm:
(WebKit::WebAutomationSession::platformSimulateKeyboardInteraction):
* Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
(WebKit::virtualKeyHasStickyModifier):
(WebKit::keyCodeForVirtualKey):
(WebKit::eventModifierFlagsForVirtualKey):
* Source/WebKit/UIProcess/Automation/gtk/WebAutomationSessionGtk.cpp:
(WebKit::keyCodeForVirtualKey):
* Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
(WebKit::keyCodeForVirtualKey):

Canonical link: <a href="https://commits.webkit.org/257406@main">https://commits.webkit.org/257406@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/926eb4639f9b2e3d6da85f512339c4a8dc90fa33

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98770 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7981 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31909 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108186 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8522 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85349 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91299 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106103 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104453 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6461 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90017 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33433 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88283 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21359 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76357 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1898 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22889 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1805 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45362 "Found 1 new test failure: media/video-inactive-playback.html (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5096 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6754 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42344 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3200 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg reviewed (approved):
LGTM!

If we are keeping test expectations in WebKit it would be great to update those at the same time. If not, no worries!

patrickangle committed:
WebDriver: [Cocoa] [Actions] [Key] Right-hand modifier keys report as their left-hand counterpart
https://bugs.webkit.org/show_bug.cgi?id=248674
rdar://100588604

Reviewed by BJ Burg.

Add support to the Cocoa ports for properly handling right-hand modifiers which, when paired with safaridriver changes,
will allow the left and right modifier keys to be distiguished from one another. Most of the support is already done,
this just adds the missing `CommandRight` virtual key to the automation protocol and adds handling so that these keys
are properly considered to be a modifier key when combined with other subsequent actions.

This progresses the following WPT test cases:
webdriver/tests/perform_actions/key_events.py::test_modifier_key_sends_correct_events[\ue052-R_ALT]
webdriver/tests/perform_actions/key_events.py::test_modifier_key_sends_correct_events[\ue051-R_CONTROL]
webdriver/tests/perform_actions/key_events.py::test_modifier_key_sends_correct_events[\ue050-R_SHIFT]
webdriver/tests/perform_actions/key_events.py::test_special_key_sends_keydown[R_ALT-expected48]
webdriver/tests/perform_actions/key_events.py::test_special_key_sends_keydown[R_CONTROL-expected53]
webdriver/tests/perform_actions/key_events.py::test_special_key_sends_keydown[R_SHIFT-expected61]

* Source/WebKit/UIProcess/Automation/Automation.json:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::normalizedVirtualKey):
* Source/WebKit/UIProcess/Automation/cocoa/WebAutomationSessionCocoa.mm:
(WebKit::WebAutomationSession::charCodeForVirtualKey const):
* Source/WebKit/UIProcess/Automation/ios/WebAutomationSessionIOS.mm:
(WebKit::WebAutomationSession::platformSimulateKeyboardInteraction):
* Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
(WebKit::virtualKeyHasStickyModifier):
(WebKit::keyCodeForVirtualKey):
(WebKit::eventModifierFlagsForVirtualKey):
* Source/WebKit/UIProcess/Automation/gtk/WebAutomationSessionGtk.cpp:
(WebKit::keyCodeForVirtualKey):
* Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionLibWPE.cpp:
(WebKit::keyCodeForVirtualKey):

Canonical link: https://commits.webkit.org/257406@main
